src/Microsoft.ML.Data/Model/Repository.cs

```diff
@@ -109,13 +109,25 @@ internal Repository(bool needDir, IExceptionContext ectx)
             _open = new List<Entry>();
             if (needDir)
             {
-                DirTemp = GetTempPath();
+                DirTemp = GetShortTempDir();
                 Directory.CreateDirectory(DirTemp);
             }
             else
                 GC.SuppressFinalize(this);
         }
 
+        private static string GetShortTempDir()
+        {
+            var rnd = RandomUtils.Create();
+            string path;
+            do
+            {
+                path = Path.Combine(Path.GetTempPath(), "TLC_" + rnd.Next().ToString("X"));
+            }
+            while (Directory.Exists(path));
+            return Path.GetFullPath(path);
```

Comment:
Back when we were using a GUID this was less likely, but now the likelihood of a collision is much higher, I start to worry a lot more about the inherent race condition in this code. :P With a 128-bit GUID, the chance of collision was on the order of 1/2^64... now it's 31 bits, it's on the order of 1/2^16.

Can we restructure this code a bit so that the *creation* of the directory and the generation of a supposedly safe directory name happen at the same time? We can handle this by creating it in a non-lazy fashion. If we want the creation to be "lazy" as it currently is, then I do not see how the currently scheme of a `protected readonly string DirTemp` can be allowed to stand.